Cleanup of init/device and kernel internal headers

nashif opened the pull request:
Started when I noticed soc code including kernel_internal.h. Then i started finding other oddities and misplaced code.

Cleanup of init.c which was getting overloaded with too many functions used by arch and soc code and remove unrelated declarations from kernel_internal.h which were forcing some non-kernel code to include kernel_internal.h


No functional changes, only code movements and header cleanup`

nashif committed:
kernel: move xip into arch/common

Not really a kernel feature, more for architecture, which is reflected
in how XIP is enabled and tested. Move it to architecture code to keep
which much of the 'implementation' and usage is.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel/arch: Move early init/boot code out of init/kernel headers

Cleanup init.c code and move early boot code into arch/ and make it
accessible outside of the boot process/kernel.

All of this code is not related to the 'kernel' and is mostly used
within the architecture boot / setup process.

The way it was done, some soc code was including kernel_internal.h
directly, which shouldn't be done.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: init: move boot arg handling to own file

No reason for this to be part of already packed init.c.
Moved to own file and build only when BOOTARGS are enabled.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
device: move device syscalls to device.c

Move device model syscalls to device.c and decouple kernel header from
device related routines. Cleanup init to have only what is needed.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: use cmake macro for adding kernel files

simplify cmake file and use macros for adding files that are part of the
kernel based on the configuration.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: xip: move under arch

Move under arch, as this is not a kernel feature really. arch also
matches the test idcentifier in place.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: gen_isr_table: move under arch

Move under arch, as this is not a kernel feature really. arch also
matches the test idcentifier in place.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: interrupt: move under arch

Move under arch, as this is not a kernel feature really. arch also
matches the test idcentifier in place.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: init: rename z_early_memset -> arch_early_memset

Do not use private API prefix and move to architecture interface as
those functions are primarily used across arches and can be defined by
the architecture.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: boot: rename z_early_memcpy -> arch_early_memcpy

Do not use private API prefix and move to architecture interface as
those functions are primarily used across arches and can be defined by
the architecture.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: init: z_bss_zero -> arch_bss_zero

Do not use private API prefix and move to architecture interface as
those functions are primarily used across arches and can be defined by
the architecture.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: init: rename z_data_copy -> arch_data_copy

Do not use private API prefix and move to architecture interface as
those functions are primarily used across arches and can be defined by
the architecture.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: boot args kconfig cleanup

Cleanup kconfig of bootargs and put everything in one menuconfig.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arch: x86: declare z_interrupt_stacks

Declare z_interrupt_stacks as extern.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
soc: espressif: declare z_cstart

Declare z_cstart as extern as otherwise we will need to pull in
kernel_internal.h.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

dcpleung commented:
Maybe time for a `boot` subsys? :D

andyross reviewed (dismissed):
Looks good here.  I particularly love the isolation of arch/common code, makes a lot of our memory setup initialization less confusing.

marekmatej commented on kernel/boot_args.c:
```diff
@@ -2,10 +2,10 @@
  * Copyright The Zephyr Project contributors
  */
 
- #include <zephyr/sys/util.h>
```
just a nit, but maybe you could move this change in the previous commit

marekmatej reviewed (dismissed):
besides a small nit looks good

cfriedt reviewed (dismissed):
LGTM 👍

peter-mitsis commented on soc/espressif/esp32/soc.c:
```diff
@@ -18,6 +18,7 @@
 
 extern void z_prep_c(void);
 extern void esp_reset_reason_init(void);
+extern FUNC_NORETURN void z_cstart(void);
```
With how things are currently structured, declaring z_cstart() here is not necessary as it is not called within this file. However, z_prep_c() is called, which in turn calls z_cstart(). This also applies to esp32s2 and esp32s3.

nashif committed:
arch: mark z_prep_c as FUNC_NORETURN

z_prep_c does not return, mark it as such consistently across
architectures.  We had some arches do that, others not. This resolves a
few coding guideline violations in arch code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
can we please get another round of reviews please? 
FYI: @dcpleung @andyross  @sylvioalves and others....

peter-mitsis commented on kernel/device.c:
```diff
@@ -27,6 +28,58 @@ void z_device_state_init(void)
+			if (rc < 0) {
+				rc = -rc;
+			}
+			if (rc > UINT8_MAX) {
+				rc = UINT8_MAX;
+			}
```
This might be beyond the scope of the work item, but if `rc` is ever INT_MIN, then -rc is still less than 0. It may be worthwhile to cast `rc` to an `unsigned int` as below to guarantee that we do not get the weird fringe case.

```
   if ((unsigned int)rc > UINT8_MAX) {
      rc = UINT8_MAX;
   }
```

nashif replied:
yeah, not part of this PR....

peter-mitsis commented on cmake/linker_script/arm/linker.cmake:
```diff
@@ -149,7 +149,7 @@ zephyr_linker_section(NAME .ramfunc GROUP RAM_REGION SUBALIGN 8)
   # This is where the app_mem_partition stuff is going to be placed, once it
   # is generated by gen_app_partitions.py. _app_smem has its own init-copy
-  # handling in z_data_copy, so put it in RAM_REGIOM rather than DATA_REGION
+  # handling in arch_data_copy, so put it in RAM_REGIOM rather than DATA_REGION
```
There is an existing typo: RAM_REGIOM -> RAM_REGION

nashif replied:
well, not part of my change, but fixed it anyways :)

peter-mitsis commented on soc/espressif/esp32c6/soc_lpcore.c:
```diff
@@ -14,6 +15,7 @@
 #include <soc.h>
 
 extern void main(void);
+extern FUNC_NORETURN void z_cstart(void);
```
Perhaps I missed it elsewhere, but what about esp32h2/soc.c ?

peter-mitsis reviewed (commented):
A couple of loose ends identified, and a couple of items that may be out of scope.
